proposals/proposaldb.go

```diff
@@ -0,0 +1,211 @@
+package proposals
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"path/filepath"
+	"sync"
+
+	"github.com/spacemeshos/go-spacemesh/codec"
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/database"
+	"github.com/spacemeshos/go-spacemesh/events"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/mesh"
+)
+
+// DB holds all data for proposals.
+type DB struct {
+	logger log.Log
+	msh    meshDB
+
+	mu        sync.Mutex
+	proposals database.Database
+	layers    database.Database
+}
+
+// dbOpt for configuring DB.
+type dbOpt func(*DB)
+
+func withProposalDB(d database.Database) dbOpt {
+	return func(db *DB) {
+		db.proposals = d
+	}
+}
+
+func withLayerDB(d database.Database) dbOpt {
+	return func(db *DB) {
+		db.layers = d
+	}
+}
+
+func withMeshDB(msh meshDB) dbOpt {
+	return func(db *DB) {
+		db.msh = msh
+	}
+}
+
+func withLogger(l log.Log) dbOpt {
+	return func(db *DB) {
+		db.logger = l
+	}
+}
+
+func newDB(opts ...dbOpt) *DB {
+	db := &DB{
+		logger: log.NewNop(),
+	}
+	for _, opt := range opts {
+		opt(db)
+	}
+	return db
+}
+
+// NewProposalDB returns a new DB for proposals.
+func NewProposalDB(path string, msh meshDB, logger log.Log) (*DB, error) {
+	pdb, err := database.NewLDBDatabase(filepath.Join(path, "proposals"), 0, 0, logger)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize proposals db: %w", err)
+	}
+	ldb, err := database.NewLDBDatabase(filepath.Join(path, "proposal_layers"), 0, 0, logger)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize proposal layers db: %w", err)
+	}
+	return newDB(withProposalDB(pdb), withLayerDB(ldb), withMeshDB(msh), withLogger(logger)), nil
+}
+
+// Close closes all resources.
+func (db *DB) Close() {
+	db.proposals.Close()
+	db.layers.Close()
+}
+
+// HasProposal returns true if the database has the Proposal specified by the ProposalID and false otherwise.
+func (db *DB) HasProposal(id types.ProposalID) bool {
+	has, err := db.proposals.Has(id.Bytes())
+	return err == nil && has
```

Comment:
I think it's not a big deal but on DB error other than the not-found error we will not know that it occurred.

Reply:
this is used in proposal handler. my consideration was, either kind of error (not found or disk error), the resulting action is similar. we still want to process the newly received proposal and save it as best effort. so checking the error wouldn't be very meaningful, and whatever disk error will recur later when the new proposal is saved.

there is similar logic for Ballot.

Reply:
I agree that such an error shouldn't change the handling process. My thought was to consider logging a disk error to know that it happened. But it's not important, so we can also leave it as is